File: Sources/Domain/AppDeepLink.swift
import Foundation

enum AppActivationTarget: String, Equatable, Sendable {
  case mainWindow
  case statistics
  case subscriptionQuota
  case controllerSettings
}

enum AppDeepLink {
  static let fallbackTarget = AppActivationTarget.mainWindow
  static let statisticsURL = "mihomo-meter://statistics"
  static let subscriptionQuotaURL = "mihomo-meter://subscription-quota"
  static let connectionSettingsURL = "mihomo-meter://connection-settings"

  static func target(for url: URL) -> AppActivationTarget? {
    switch url.absoluteString {
    case statisticsURL:
      .statistics
    case subscriptionQuotaURL:
      .subscriptionQuota
    case connectionSettingsURL:
      .controllerSettings
    default:
      nil
    }
  }
}

File: platform/windows/MihomoMeter.Windows.Core/Domain/AppDeepLink.cs
namespace MihomoMeter.Windows.Core.Domain;

public enum AppActivationTarget
{
    MainWindow,
    Statistics,
    SubscriptionQuota,
    ControllerSettings,
}

public static class AppActivationTargetContract
{
    public static string Value(AppActivationTarget target)
    {
        return target switch
        {
            AppActivationTarget.MainWindow => "main",
            AppActivationTarget.Statistics => "statistics",
            AppActivationTarget.SubscriptionQuota => "subscriptionQuota",
            AppActivationTarget.ControllerSettings => "controllerSettings",
            _ => throw new ArgumentOutOfRangeException(nameof(target)),
        };
    }

    public static bool TryParse(string? value, out AppActivationTarget target)
    {
        target = value switch
        {
            "main" => AppActivationTarget.MainWindow,
            "statistics" => AppActivationTarget.Statistics,
            "subscriptionQuota" => AppActivationTarget.SubscriptionQuota,
            "controllerSettings" => AppActivationTarget.ControllerSettings,
            _ => default,
        };
        return value is "main" or "statistics" or "subscriptionQuota" or "controllerSettings";
    }
}

public static class AppDeepLink
{
    public const string StatisticsUrl = "mihomo-meter://statistics";
    public const string SubscriptionQuotaUrl = "mihomo-meter://subscription-quota";
    public const string ConnectionSettingsUrl = "mihomo-meter://connection-settings";

    public static bool TryParse(Uri? uri, out AppActivationTarget target)
    {
        target = uri?.OriginalString switch
        {
            StatisticsUrl => AppActivationTarget.Statistics,
            SubscriptionQuotaUrl => AppActivationTarget.SubscriptionQuota,
            ConnectionSettingsUrl => AppActivationTarget.ControllerSettings,
            _ => AppActivationTarget.MainWindow,
        };
        return uri?.OriginalString is StatisticsUrl
            or SubscriptionQuotaUrl
            or ConnectionSettingsUrl;
    }
}
